src/servers/src/postgres/types.rs

```diff
@@ -0,0 +1,677 @@
+        &Type::VARCHAR | &Type::TEXT => Ok(format!(
+            "'{}'",
+            portal.parameter::<String>(idx)?.as_deref().unwrap_or("")
+        )),
+        &Type::BOOL => Ok(portal
+            .parameter::<bool>(idx)?
+            .map(|v| v.to_string())
+            .unwrap_or_else(|| "".to_owned())),
+        &Type::INT4 => Ok(portal
+            .parameter::<i32>(idx)?
+            .map(|v| v.to_string())
+            .unwrap_or_else(|| "".to_owned())),
+        &Type::INT8 => Ok(portal
+            .parameter::<i64>(idx)?
+            .map(|v| v.to_string())
+            .unwrap_or_else(|| "".to_owned())),
+        &Type::FLOAT4 => Ok(portal
+            .parameter::<f32>(idx)?
+            .map(|v| v.to_string())
+            .unwrap_or_else(|| "".to_owned())),
+        &Type::FLOAT8 => Ok(portal
+            .parameter::<f64>(idx)?
+            .map(|v| v.to_string())
+            .unwrap_or_else(|| "".to_owned())),
+        &Type::TIMESTAMP => Ok(portal
+            .parameter::<NaiveDateTime>(idx)?
+            .map(|v| v.format("%Y-%m-%d %H:%M:%S%.6f").to_string())
+            .unwrap_or_else(|| "".to_owned())),
+        _ => Err(PgWireError::UserError(Box::new(ErrorInfo::new(
+            "ERROR".to_owned(),
+            "22023".to_owned(),
+            "unsupported_parameter_value".to_owned(),
+        )))),
+    }
+}
+
+pub(super) fn invalid_parameter_error(msg: &str, detail: Option<&str>) -> PgWireError {
+    let mut error_info = ErrorInfo::new("ERROR".to_owned(), "22023".to_owned(), msg.to_owned());
+    error_info.set_detail(detail.map(|s| s.to_owned()));
+    PgWireError::UserError(Box::new(error_info))
+}
+
+pub(super) fn parameters_to_scalar_values(
+    plan: &LogicalPlan,
+    portal: &Portal<SqlPlan>,
+) -> PgWireResult<Vec<ScalarValue>> {
+    let param_count = portal.parameter_len();
+    let mut results = Vec::with_capacity(param_count);
+
+    let client_param_types = portal.statement().parameter_types();
+    let param_types = plan
+        .get_param_types()
+        .map_err(|e| PgWireError::ApiError(Box::new(e)))?;
+
+    // ensure parameter count consistent for: client parameter types, server
+    // parameter types and parameter count
+    if param_types.len() != param_count {
+        return Err(invalid_parameter_error(
+            "invalid_parameter_count",
+            Some(&format!(
+                "Expected: {}, found: {}",
+                param_types.len(),
+                param_count
+            )),
+        ));
+    }
+    if client_param_types.len() != param_count {
+        return Err(invalid_parameter_error(
+            "invalid_parameter_count",
+            Some(&format!(
+                "Expected: {}, found: {}",
+                client_param_types.len(),
+                param_count
+            )),
+        ));
+    }
+
+    for (idx, client_type) in client_param_types.iter().enumerate() {
+        if let Some(Some(server_type)) = param_types.get(&format!("${}", idx + 1)) {
```

Comment:
```suggestion
        let Some(Some(server_type)) = param_types.get(&format!("${}", idx + 1)) else { continue };
```
to reduce a layer of indentation